src/internals/ir/callGraph.ts

```diff
@@ -0,0 +1,196 @@
+import { unreachable } from "../util";
+import { TactASTStore } from "./astStore";
+import { IdxGenerator } from "./indices";
+import { MistiContext } from "../../";
+import { Logger } from "../../internals/logger";
+import { forEachExpression } from "../tact/iterators";
+import {
+  AstFunctionDef,
+  AstReceiver,
+  AstContractInit,
+  AstExpression,
+  AstMethodCall,
+  AstStaticCall,
+} from "@tact-lang/compiler/dist/grammar/ast";
+
+type CGNodeId = number & { readonly brand: unique symbol };
+type CGEdgeId = number & { readonly brand: unique symbol };
+
+/**
+ * Represents an edge in the call graph, indicating a call from one function to another.
+ */
+class CGEdge {
+  public idx: CGEdgeId;
+
+  constructor(
+    public src: CGNodeId,
+    public dst: CGNodeId,
+  ) {
+    this.idx = IdxGenerator.next("cg_edge") as CGEdgeId;
+  }
+}
+
+/**
+ * Represents a node in the call graph, corresponding to a function or method.
+ */
+class CGNode {
+  public idx: CGNodeId;
+  public inEdges: Set<CGEdgeId> = new Set();
+  public outEdges: Set<CGEdgeId> = new Set();
+
+  constructor(
+    public astId: number | undefined,
+    public name: string,
+  ) {
+    this.idx = IdxGenerator.next("cg_node") as CGNodeId;
+  }
+}
+
+export class CallGraph {
+  private nodeMap: Map<CGNodeId, CGNode> = new Map();
+  private edgesMap: Map<CGEdgeId, CGEdge> = new Map();
+  private nameToNodeId: Map<string, CGNodeId> = new Map();
+  private logger: Logger;
+
+  constructor(private ctx: MistiContext) {
+    this.logger = ctx.logger;
+  }
+
+  public getNodes(): Map<CGNodeId, CGNode> {
+    return this.nodeMap;
+  }
+
+  public getEdges(): Map<CGEdgeId, CGEdge> {
+    return this.edgesMap;
+  }
+
+  public build(astStore: TactASTStore): CallGraph {
+    for (const func of astStore.getFunctions()) {
+      const funcName = this.getFunctionName(func);
+      if (funcName) {
+        const node = new CGNode(func.id, funcName);
+        this.nodeMap.set(node.idx, node);
+        this.nameToNodeId.set(funcName, node.idx);
+      } else {
+        this.logger.error(
+          `Function with id ${func.id} has no name and will be skipped.`,
+        );
+      }
+    }
+    this.analyzeFunctionCalls(astStore);
+    return this;
+  }
+
+  public areConnected(src: CGNodeId, dst: CGNodeId): boolean {
```

Comment:
That's a part of the public API we provide to users, thus, it should be documented with a docstring.